pandas/core/groupby/indexing.py

```diff
@@ -0,0 +1,238 @@
+        
+        Used to implement GroupBy._rows which is used to implement GroupBy.nth
+        when keyword dropna is None or absent.
+        The behaviour extends GroupBy.nth and handles DataFrame.groupby() 
+        keyword parameters such as as_index and dropna in a compatible way.
+        
+        The additions to nth(arg) are:
+        - Handles iterables such as range.
+        - Handles slice(start, stop, step) with
+            start: positive, negative or None.
+            stop: positive, negative or None.
+            step: positive or None.
+        
+        Parameters
+        ----------
+        arg : PositionalIndexer | tuple
+            Allowed values are:
+            - Integer
+            - Integer values iterable such as list or range
+            - Slice
+            - Comma separated list of integers and slices
+        
+        Returns
+        -------
+        Series
+            The filtered subset of the original groupby Series.
+        DataFrame
+            The filtered subset of the original groupby DataFrame.
+
+        See Also
+        --------
+        DataFrame.iloc : Purely integer-location based indexing for selection by
+            position.
+        GroupBy.head : Return first n rows of each group.
+        GroupBy.tail : Return last n rows of each group.
+        GroupBy.nth : Take the nth row from each group if n is an int, or a
+            subset of rows, if n is a list of ints.
+
+        Examples
+        --------
+            >>> df = pd.DataFrame([["a", 1], ["a", 2], ["a", 3], ["b", 4], ["b", 5]],
+            ...                   columns=["A", "B"])
+            >>> df.groupby("A", as_index=False)._rows[1:2]
+               A  B
+            1  a  2
+            4  b  5
+
+            >>> df.groupby("A", as_index=False)._rows[1, -1]
+               A  B
+            1  a  2
+            2  a  3
+            4  b  5
+        """
+        with groupby.group_selection_context(self.grouped):
+            if isinstance(arg, tuple):
+                if all(is_integer(i) for i in arg):
+                    mask = self._handle_list(arg)
+
+                else:
+                    mask = self._handle_tuple(arg)
+
+            elif isinstance(arg, slice):
+                mask = self._handle_slice(arg)
+
+            elif is_integer(arg):
+                mask = self._handle_int(cast(int, arg))
+
+            elif is_list_like(arg):
+                mask = self._handle_list(cast(Iterable[int], arg))
+
+            else:
+                raise TypeError(
+                    f"Invalid index {type(arg)}. "
+                    "Must be integer, list-like, slice or a tuple of "
+                    "integers and slices"
+                )
+
+            ids, _, _ = self.grouped.grouper.group_info
+
+            # Drop NA values in grouping
+            mask &= ids != -1
+
+            if mask is None or mask is True:
+                result = self.grouped._selected_obj[:]
```

Comment:
does _selected_obj vs. _obj_with_exclusions make a difference here?

Reply:
I don't know. This is the method used by head, tail and nth.

Reply:
> does _selected_obj vs. _obj_with_exclusions make a difference here?

There is no `_obj_with_exclusions` on groupby objects. `_obj_with_exclusions` and `_selected_obj` are defined on frames and series, but just `_selected_obj` for groupby.

Reply:
```
df = pd.DataFrame({'a': [1, 2], 'b': [3, 4]})
print(df.groupby('a')._obj_with_exclusions)
```

produces

```
   b
0  3
1  4
```

Reply:
You are right. It comes in from SelectionMixin and seems to lose the grouped column.
But I think that _selected_obj gives the best behaviour. Especially to emulate head and tail, that essentially give subsets of the original df.